lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,471 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This file includes methods to manage TLS certificates within the Vault charms."""
+
+import logging
+import socket
+from enum import Enum, auto
+from signal import SIGHUP
+from typing import List, Optional, TextIO, Tuple
+
+from charms.certificate_transfer_interface.v0.certificate_transfer import (
+    CertificateTransferProvides,
+)
+from charms.tls_certificates_interface.v2.tls_certificates import (
+    TLSCertificatesRequiresV2,
+    csr_matches_certificate,
+    generate_ca,
+    generate_certificate,
+    generate_csr,
+    generate_private_key,
+)
+from ops import EventBase, Object, RelationBrokenEvent, SecretNotFoundError
+from ops.pebble import APIError, PathError
+
+from exceptions import PeerSecretError, VaultCertsError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "61b41a053d9847ce8a14eb02197d12cb"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+SEND_CA_CERT_RELATION_NAME = "send-ca-cert"
+TLS_CERTIFICATE_ACCESS_RELATION_NAME = "tls-certificates-access"
+TLS_CERT_FILE_PATH = "/vault/certs/cert.pem"
+TLS_KEY_FILE_PATH = "/vault/certs/key.pem"
+TLS_CA_FILE_PATH = "/vault/certs/ca.pem"
+TLS_CSR_FILE_PATH = "/vault/certs/csr.pem"
+TLS_FILE_FOLDER_PATH = "/vault/certs/"
+CA_CERTIFICATE_JUJU_SECRET_KEY = "vault-ca-certificates-secret-id"
+CA_CERTIFICATE_JUJU_SECRET_LABEL = "vault-ca-certificate"
+
+VAULT_CA_SUBJECT = "Vault self signed CA"
+
+
+class Substrate(Enum):
+    """https://juju.is/docs/juju/cloud TODO: explain."""
```

Comment:
Can you add some details?

Reply:
Done. For more context, the way I see it is we have 2 main ways of reusing this library:

1) We explicitly pass if we're in K8s or machine charm so we can use conditionals to pick what we're going to do.
2) We can pass the container or the machine objects that share an abstract base class, and just call `pull`, `push` etc. on this class. 

I saw that @gruyaume already put a Machine class in the machine charm of Vault, but I'm still wondering if we should have a discussion about this?